api/proto/config/storage.proto

```diff
@@ -242,3 +242,32 @@ message DisksConfig {
   DisksArrayType array_type = 2;
   repeated DisksConfig children = 3;
 }
+
+
+// SnapshotConfig describes a snapshot for a specific logical
+// volume that must exist on the device. It has a required field
+// volume_uuid for which it was created and UUID for snapshot name in ZFS.
+message SnapshotConfig {
+  // The real name of the snapshot in ZFS. It is the link between
+  // the command and the response to the command. It is assumed
+  // that the field will always be filled with a unique value
+  // that the controller will generate.
+  string uuid = 1;
+  // Display name (User-friendly name). This name does not affect
+  // the snapshot properties in ZFS in any way.
+  // Can be filled on the cloud side.
+  // If a snapshot has already been created and this field has changed,
+  // it can be assumed that the friendly name for this snapshot has
+  // been renamed on the cloud side.
+  string display_name = 2;
+  // Volume ID of the volume this snapshot belongs to. The field is
+  // required for all messages. Must always be filled in on the
+  // cloud side before sending to create snapshot command.
+  string volume_uuid = 3;
+  // The command for the snapshot rollback operation.
+  // Keep in mind that in order to return to any of the
+  // previous snapshots, all intermediate snapshots will be
+  // destroyed in ZFS. These snapshots must be marked as
+  // deleted in the SnapshotInfo message.
```

Comment:
Should that say SnapshotConfig?

Reply:
Or is the assumption that EVE-OS delete the newer snapshots implicitly as a result of the rollback?
Technically couldn't we keep them around (to allow a roll forward?) and have the controller explicitly remove any snapshots it wants deleted from the SnapshotConfig?

Reply:
I thought it was worth mentioning here

The fact is that by default in ZFS, a rollback to the last snapshot is implied. But this does not mean that we cannot roll back to earlier snapshots. In this case, ZFS in EVE implicitly automatically deletes all snapshots taken after the snapshot to which we rolled back. You can read about it [here](https://docs.oracle.com/cd/E19253-01/819-5461/gbcxk/index.html).
In the logic of EVE, after rollback ops, we check all snapshots and mark the missing snapshots in ZFS as deleted, and send this information to the controller, expecting that configurations with them will no longer come from the controller.

There is another option. The controller can take this fact into account and not send configurations for snapshots made after the snapshot to which we rolled back. Thus, in EVE we will consider them remote. And we will also send a message that they have been deleted.

Reply:
I think this should be captured as a comment somewhere. Maybe doocumented in the API proto file where you can suggest that when the controller says to rollback to snapshot k that it should also delete k+1 through n.

Also, I think we want to make a distinction in the info message between snapshots deleted because the controller asked us to delete them, and this case of implicitly forcibly deleted as a side effect of the rollback.
Typically how we handle a delete of other objects from the controller is 1) to remove it from the EdgeDevConfig and 2) no longer reporting it in the info message. If we take the same approach here it means that the Deleted in the info message would only be used in the case when it is implicitly deleted by ZFS due to e.g., a rollback. (Maybe there will be other implicit deletes if we have resource limits in EVE-OS where we limit the amount of storage used for the snapshots where this IMPLICITLY_DELETED would make sense down the road.)

So maybe the Info DELETED should be commented and/or renamed to IMPLICITLY_DELETED?

Reply:
> I think this should be captured as a comment somewhere. Maybe doocumented in the API proto file where you can suggest that when the controller says to rollback to snapshot k that it should also delete k+1 through n.

Ok

> Typically how we handle a delete of other objects from the controller is 1) to remove it from the EdgeDevConfig and 2) no longer report it in the info message. If we take the same approach here it means that the Deleted in the info message would only be used in the case when it is implicitly deleted by ZFS due to e.g., a rollback. (Maybe there will be other implicit deletes if we have resource limits in EVE-OS where we limit the amount of storage used for the snapshots where this IMPLICITLY_DELETED would make sense down the road.)

Here I agree with you since we really do not send information that the snapshot was deleted in a situation where the configuration for it did not come to us and it was successfully deleted the first time.
But in this case, we can send information if, for some unknown reason, we were not able to delete the snapshot the first time. Then we fill in the error variable, send the "DELETING" status to the controller, and after a short time, we make another deletion attempt. If N the deletion attempt was successful, then we mark the snapshot as deleted, clean up the error, and report the successful deletion to the controller.

**Question:** _Should we report a successful deletion after the "DELETING" status, or should we do the same as if we deleted the snapshot on the first try?_

I implemented this logic based on the fact that a situation is possible when we send a command from the controller to delete the snapshot (for example, in order to make room for the next application). And in this case, if you do not report problems during the execution of the delete command and its successful execution, a situation may occur when the snapshot in the controller was deleted, the space for the application space was not freed up and the application was not created. Of course, the controller will report that the space has run out and it is impossible to create a new logical volume, but then the user may be indignant since he freed up space a couple of steps back by deleting the snapshot.

I can add a new status IMPLICITLY_DELETED which will be set based on when the current snapshot state is CREATED but the snapshot itself is not in ZFS for reasons unknown to us.

In addition, we have other implicit snapshot deletions, such as deleting a logical volume.


> So maybe the Info DELETED should be commented and/or renamed to IMPLICITLY_DELETED?

I will add a comment.
The very situation about the need for the IMPLICITLY_DELETED status boils down to the question of whether the controller will process the deletion of snapshots from k + 1 to n when rolling back to snapshot k. And delete snapshots before deleting the logical volume. But I believe that we no should rely on the logic of the controller, so I propose to extend the statuses by adding IMPLICITLY_DELETED if we need to report the successful removal of the snapshot after unsuccessful attempts. If this is not required then I can rename DELETED to IMPLICITLY_DELETED.

Reply:
The controller wouldn't be sending some delete command but instead no longer include the particular snapshot in edgedevconfig. Until that is successful the snapshot will be reported by EVE-OS in the info message. The entry in the info message would contain a DELETING message if the operation takes some time, and some errorinfo if the delete fails.
Thus we don't need a DELETED in the info message; just an IMPLICITLY_DELETED.

Reply:
Ok, i corrected it.